src/java/org/apache/cassandra/db/virtual/AccordVirtualTables.java

```diff
@@ -246,6 +259,106 @@ public DataSet data()
+                                Arrays.asList(UTF8Type.instance, UTF8Type.instance), false);
+        }
+
+        private ByteBuffer pk(PartitionKey pk)
+        {
+            var tm = Schema.instance.getTableMetadata(pk.table());
+            return partitionKeyType.pack(UTF8Type.instance.decompose(tm.toString()),
+                                         UTF8Type.instance.decompose(pk.token().toString()));
+        }
+
+        @Override
+        public Iterable<UserType> userTypes()
+        {
+            return Arrays.asList(partitionKeyType);
+        }
+
+        @Override
+        public DataSet data(DecoratedKey partitionKey)
+        {
+            TxnId id = TxnId.parse(UTF8Type.instance.compose(partitionKey.getKey()));
+            List<StoreTxnState> shards = AccordService.instance().debug(id);
+
+            SimpleDataSet ds = new SimpleDataSet(metadata());
+            for (StoreTxnState shard : shards)
+            {
+                Set<TxnId> processed = new HashSet<>();
+                process(ds, shard, processed, id, 0, id, Reason.Self, null);
+                // everything was processed right?
+                Sets.SetView<TxnId> skipped = Sets.difference(shard.txns.keySet(), processed);
+                if (!skipped.isEmpty())
+                    throw new IllegalStateException("Skipped txns: " + skipped);
+            }
+
+            return ds;
+        }
+
+        private void process(SimpleDataSet ds, StoreTxnState shard, Set<TxnId> processed, TxnId userTxn, int depth, TxnId txnId, Reason reason, Runnable onDone)
+        {
+            if (!processed.add(txnId))
+                throw new IllegalStateException("Double processed " + txnId);
+            StoreTxnState.TxnState txn = shard.txns.get(txnId);
+            if (txn == null)
+            {
+                Invariants.checkState(reason == Reason.Self, "Txn %s unknown for reason %s", txnId, reason);
+                return;
+            }
+            ds.row(userTxn.toString(), shard.storeId, depth, reason == Reason.Self ? "" : txn.txnId.toString(), reason.name());
+            ds.column("save_status", txn.saveStatus.name());
+            if (txn.executeAt != null)
+                ds.column("execute_at", txn.executeAt.toString());
+            if (onDone != null)
+                onDone.run();
+            if (!txn.notBlocked())
+            {
+                for (TxnId blockedBy : txn.blockedBy)
+                {
+                    if (processed.contains(blockedBy)) continue; // already listed
+                    process(ds, shard, processed, userTxn, depth + 1, blockedBy, Reason.Txn, null);
+                }
+                for (PartitionKey blockedBy : txn.blockedByKey)
+                {
+                    TxnId blocking = shard.keys.get(blockedBy);
+                    if (processed.contains(blocking)) continue; // already listed
```

Comment:
Why do we need the `processed` set?

The only possible loop we should be able to hit would be if we are exploring what is "blocking" an uncommitted transaction, but this would be a mistake - the only thing blocking an uncommitted transaction is itself. Once we hit an uncommitted transaction we should stop exploring further.

I'm not opposed to keeping the collection as a safety mechanism, and using it to return a special warning if we detect a loop, but this would be a serious bug.

Reply:
Oh, is this in part because we're exploring via potentially multiple routes, rather than just the actually currently blocking route?

Reply:
> The only possible loop

this is not only due to loops.  Lets say we have the following

```
T1
T2 blocked by T1
T3 blocked by T1
T4 blocked by key K1 (nextApply=T1)
```

Can I say that T3 will only ever depend on T2? don't know... can I say that some txn will depend directly while others depend on keys?  don't know... mostly figure its safer to keep track than to fail in a random way if/when this happens.